crates/ruff_cli/src/cache.rs

```diff
@@ -1,171 +1,238 @@
+            .with_context(|| format!("Failed parse cache file '{}'", path.display()))?;
 
-        let files = serialize_messages.files.take();
-
-        s.serialize_field("files", &files)?;
+        // Sanity check.
+        if cache.package_root != package_root {
+            return Err(anyhow!(
+                "Different package root in cache: expected '{}', got '{}'",
+                package_root.display(),
+                cache.package_root.display(),
+            ));
+        }
 
-        s.end()
+        cache.path = path;
+        Ok(cache)
     }
-}
 
-struct SerializeMessages<'a> {
-    messages: &'a [Message],
-    files: RefCell<Files<'a>>,
-}
-
-impl Serialize for SerializeMessages<'_> {
-    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        let mut s = serializer.serialize_seq(Some(self.messages.len()))?;
-        let mut files = self.files.borrow_mut();
+    /// Store the cache to disk.
+    pub(crate) fn store(&self) -> Result<()> {
+        let file = File::create(&self.path)
+            .with_context(|| format!("Failed to create cache file '{}'", self.path.display()))?;
+        let writer = BufWriter::new(file);
+        bincode::serialize_into(writer, &self).with_context(|| {
+            format!(
+                "Failed to serialise cache to file '{}'",
+                self.path.display()
+            )
+        })
+    }
 
-        for message in self.messages {
-            // Using a Vec instead of a HashMap because the cache is per file and the large majority of
-            // files have exactly one source file.
-            let file_id = if let Some(position) = files
-                .iter()
-                .position(|(filename, _)| *filename == message.filename())
-            {
-                position
-            } else {
-                let index = files.len();
-                files.push((message.filename(), message.file.source_text()));
-                index
-            };
+    /// Returns the relative path based on `path` and the package root.
+    ///
+    /// Returns `None` if `path` is not within the package.
+    pub(crate) fn relative_path<'a>(&self, path: &'a Path) -> Option<&'a RelativePath> {
+        path.strip_prefix(&self.package_root).ok()
+    }
 
-            s.serialize_element(&SerializeMessage { message, file_id })?;
+    /// Get the cached results for a single file at relative `path`. This uses
+    /// `file_last_modified` to determine if the results are still accurate
+    /// (i.e. if the file hasn't been modified since the cached run).
+    ///
+    /// This returns `None` if `file_last_modified` differs from the cached
+    /// timestamp or if the cache doesn't contain results for the file.
+    pub(crate) fn get(
+        &self,
+        path: &RelativePath,
+        file_last_modified: SystemTime,
+    ) -> Option<FileCache> {
+        let file = self.files.lock().unwrap().get(path)?.clone();
```

Comment:
Does this unnecessarily clone in the `file.last_modified != file_last_modified` case?

Reply:
Yes, I tried to reduce the locking time (initially I had some more code in here), but a single compare in the lock should be fine.